File: lib/models/spoken_language.dart
import 'package:flutter/material.dart';
import 'package:ui_components/components/icon/flag_icon.dart';
import '../l10n/app_localizations.dart';

class SpokenLanguage {
  final String code;

  const SpokenLanguage({required this.code});

  /// Get the country code for this language's flag
  String get countryCode => SupportedLanguages.countryCodeFor(code);

  /// Get the FlagIcon widget for this language
  Widget getFlagIcon({double size = 24.0}) {
    return FlagIcon(countryCode, size: size);
  }

  /// Get the localized name for this language
  String getName(BuildContext context) {
    final loc = AppLocalizations.of(context)!;
    return SupportedLanguages.localizedNameFor(code, loc);
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is SpokenLanguage && other.code == code;
  }

  @override
  int get hashCode => code.hashCode;

  @override
  String toString() => 'SpokenLanguage(code: $code)';
}

class _LanguageInfo {
  final String countryCode;
  final String Function(AppLocalizations) nameSelector;

  const _LanguageInfo({required this.countryCode, required this.nameSelector});
}

class SupportedLanguages {
  static final Map<String, _LanguageInfo> _languageData = {
    'en': _LanguageInfo(
      countryCode: 'united_states',
      nameSelector: (loc) => loc.languageEnglish,
    ),
    'zh': _LanguageInfo(
      countryCode: 'china',
      nameSelector: (loc) => loc.languageChinese,
    ),
    'de': _LanguageInfo(
      countryCode: 'germany',
      nameSelector: (loc) => loc.languageGerman,
    ),
    'es': _LanguageInfo(
      countryCode: 'spain',
      nameSelector: (loc) => loc.languageSpanish,
    ),
    'ru': _LanguageInfo(
      countryCode: 'russia',
      nameSelector: (loc) => loc.languageRussian,
    ),
    'ko': _LanguageInfo(
      countryCode: 'south_korea',
      nameSelector: (loc) => loc.languageKorean,
    ),
    'fr': _LanguageInfo(
      countryCode: 'france',
      nameSelector: (loc) => loc.languageFrench,
    ),
    'ja': _LanguageInfo(
      countryCode: 'japan',
      nameSelector: (loc) => loc.languageJapanese,
    ),
    'pt': _LanguageInfo(
      countryCode: 'portugal',
      nameSelector: (loc) => loc.languagePortuguese,
    ),
    'tr': _LanguageInfo(
      countryCode: 'turkey',
      nameSelector: (loc) => loc.languageTurkish,
    ),
    'pl': _LanguageInfo(
      countryCode: 'poland',
      nameSelector: (loc) => loc.languagePolish,
    ),
    'nl': _LanguageInfo(
      countryCode: 'netherlands',
      nameSelector: (loc) => loc.languageDutch,
    ),
    'ar': _LanguageInfo(
      countryCode: 'saudi_arabia',
      nameSelector: (loc) => loc.languageArabic,
    ), // Arabic uses Saudi Arabia flag
    'sv': _LanguageInfo(
      countryCode: 'sweden',
      nameSelector: (loc) => loc.languageSwedish,
    ),
    'it': _LanguageInfo(
      countryCode: 'italy',
      nameSelector: (loc) => loc.languageItalian,
    ),
    'id': _LanguageInfo(
      countryCode: 'indonesia',
      nameSelector: (loc) => loc.languageIndonesian,
    ),
    'fi': _LanguageInfo(
      countryCode: 'finland',
      nameSelector: (loc) => loc.languageFinnish,
    ),
    'vi': _LanguageInfo(
      countryCode: 'vietnam',
      nameSelector: (loc) => loc.languageVietnamese,
    ),
    'he': _LanguageInfo(
      countryCode: 'israel',
      nameSelector: (loc) => loc.languageHebrew,
    ),
    'uk': _LanguageInfo(
      countryCode: 'ukraine',
      nameSelector: (loc) => loc.languageUkrainian,
    ),
    'el': _LanguageInfo(
      countryCode: 'greece',
      nameSelector: (loc) => loc.languageGreek,
    ),
    'ms': _LanguageInfo(
      countryCode: 'malaysia',
      nameSelector: (loc) => loc.languageMalay,
    ),
    'cs': _LanguageInfo(
      countryCode: 'czech_republic',
      nameSelector: (loc) => loc.languageCzech,
    ),
    'ro': _LanguageInfo(
      countryCode: 'romania',
      nameSelector: (loc) => loc.languageRomanian,
    ),
    'da': _LanguageInfo(
      countryCode: 'denmark',
      nameSelector: (loc) => loc.languageDanish,
    ),
    'hu': _LanguageInfo(
      countryCode: 'hungary',
      nameSelector: (loc) => loc.languageHungarian,
    ),
    'no': _LanguageInfo(
      countryCode: 'norway',
      nameSelector: (loc) => loc.languageNorwegian,
    ),
    'hr': _LanguageInfo(
      countryCode: 'croatia',
      nameSelector: (loc) => loc.languageCroatian,
    ),
    'bg': _LanguageInfo(
      countryCode: 'bulgaria',
      nameSelector: (loc) => loc.languageBulgarian,
    ),
    'lt': _LanguageInfo(
      countryCode: 'lithuania',
      nameSelector: (loc) => loc.languageLithuanian,
    ),
    'la': _LanguageInfo(
      countryCode: 'vatican_city',
      nameSelector: (loc) => loc.languageLatin,
    ), // Latin uses Vatican City flag
    'mi': _LanguageInfo(
      countryCode: 'new_zealand',
      nameSelector: (loc) => loc.languageMaori,
    ), // Maori uses New Zealand flag
    'sk': _LanguageInfo(
      countryCode: 'slovakia',
      nameSelector: (loc) => loc.languageSlovak,
    ),
    'te': _LanguageInfo(
      countryCode: 'india',
      nameSelector: (loc) => loc.languageTelugu,
    ), // Telugu uses India flag
    'fa': _LanguageInfo(
      countryCode: 'iran',
      nameSelector: (loc) => loc.languagePersian,
    ),
    'lv': _LanguageInfo(
      countryCode: 'latvia',
      nameSelector: (loc) => loc.languageLatvian,
    ),
    'bn': _LanguageInfo(
      countryCode: 'bangladesh',
      nameSelector: (loc) => loc.languageBengali,
    ),
    'sr': _LanguageInfo(
      countryCode: 'serbia',
      nameSelector: (loc) => loc.languageSerbian,
    ),
    'sl': _LanguageInfo(
      countryCode: 'slovenia',
      nameSelector: (loc) => loc.languageSlovenian,
    ),
    'kn': _LanguageInfo(
      countryCode: 'india',
      nameSelector: (loc) => loc.languageKannada,
    ), // Kannada uses India flag
    'et': _LanguageInfo(
      countryCode: 'estonia',
      nameSelector: (loc) => loc.languageEstonian,
    ),
    'mk': _LanguageInfo(
      countryCode: 'republic_of_macedonia',
      nameSelector: (loc) => loc.languageMacedonian,
    ),
    'eu': _LanguageInfo(
      countryCode: 'spain',
      nameSelector: (loc) => loc.languageBasque,
    ), // Basque uses Spain flag
    'hy': _LanguageInfo(
      countryCode: 'armenia',
      nameSelector: (loc) => loc.languageArmenian,
    ),
    'mn': _LanguageInfo(
      countryCode: 'mongolia',
      nameSelector: (loc) => loc.languageMongolian,
    ),
    'bs': _LanguageInfo(
      countryCode: 'bosnia_and_herzegovina',
      nameSelector: (loc) => loc.languageBosnian,
    ),
    'kk': _LanguageInfo(
      countryCode: 'kazakhstan',
      nameSelector: (loc) => loc.languageKazakh,
    ),
    'gl': _LanguageInfo(
      countryCode: 'spain',
      nameSelector: (loc) => loc.languageGalician,
    ), // Galician uses Spain flag
    'sn': _LanguageInfo(
      countryCode: 'zimbabwe',
      nameSelector: (loc) => loc.languageShona,
    ),
    'yo': _LanguageInfo(
      countryCode: 'nigeria',
      nameSelector: (loc) => loc.languageYoruba,
    ), // Yoruba uses Nigeria flag
    'so': _LanguageInfo(
      countryCode: 'somalia',
      nameSelector: (loc) => loc.languageSomali,
    ),
    'af': _LanguageInfo(
      countryCode: 'south_africa',
      nameSelector: (loc) => loc.languageAfrikaans,
    ),
    'oc': _LanguageInfo(
      countryCode: 'france',
      nameSelector: (loc) => loc.languageOccitan,
    ), // Occitan uses France flag
    'ka': _LanguageInfo(
      countryCode: 'georgia',
      nameSelector: (loc) => loc.languageGeorgian,
    ),
    'be': _LanguageInfo(
      countryCode: 'belarus',
      nameSelector: (loc) => loc.languageBelarusian,
    ),
    'tg': _LanguageInfo(
      countryCode: 'tajikistan',
      nameSelector: (loc) => loc.languageTajik,
    ),
    'lo': _LanguageInfo(
      countryCode: 'laos',
      nameSelector: (loc) => loc.languageLao,
    ),
    'uz': _LanguageInfo(
      countryCode: 'uzbekistan',
      nameSelector: (loc) => loc.languageUzbek,
    ),
    'mt': _LanguageInfo(
      countryCode: 'malta',
      nameSelector: (loc) => loc.languageMaltese,
    ),
    'sa': _LanguageInfo(
      countryCode: 'india',
      nameSelector: (loc) => loc.languageSanskrit,
    ), // Sanskrit uses India flag
    'mg': _LanguageInfo(
      countryCode: 'madagascar',
      nameSelector: (loc) => loc.languageMalagasy,
    ),
    'as': _LanguageInfo(
      countryCode: 'india',
      nameSelector: (loc) => loc.languageAssamese,
    ), // Assamese uses India flag
    'tt': _LanguageInfo(
      countryCode: 'russia',
      nameSelector: (loc) => loc.languageTatar,
    ), // Tatar uses Russia flag
    'ln': _LanguageInfo(
      countryCode: 'democratic_republic_of_congo',
      nameSelector: (loc) => loc.languageLingala,
    ), // Lingala uses DRC flag
    'ba': _LanguageInfo(
      countryCode: 'russia',
      nameSelector: (loc) => loc.languageBashkir,
    ), // Bashkir uses Russia flag
  };

  static List<SpokenLanguage> get all =>
      _languageData.keys.map((code) => SpokenLanguage(code: code)).toList();

  static SpokenLanguage get defaultLanguage => SpokenLanguage(code: 'en');

  static SpokenLanguage? findByCode(String code) {
    return _languageData.containsKey(code) ? SpokenLanguage(code: code) : null;
  }

  static String countryCodeFor(String code) {
    final info = _languageData[code];
    return info?.countryCode ?? 'united_states';
  }

  static String localizedNameFor(String code, AppLocalizations loc) {
    final info = _languageData[code];
    return info?.nameSelector(loc) ?? code;
  }
}

File: models/SpokenLanguage.ts
export interface SpokenLanguage {
  code: string;
  name: string;
}

export interface TranscribeOptions {
  language?: string;
  prompt?: string;
}

interface LanguageInfo {
  countryCode: string;
  name: string;
  whisperLanguage?: string;
  prompt?: string;
}

const languageData: Record<string, LanguageInfo> = {
  en: { countryCode: "united_states", name: "English" },
  zh: {
    countryCode: "china",
    name: "Chinese (简体)",
    prompt: "以下是普通话的句子。",
  },
  "zh-hant": {
    countryCode: "taiwan",
    name: "Chinese (繁體)",
    whisperLanguage: "zh",
    prompt: "以下是普通話的句子。",
  },
  de: { countryCode: "germany", name: "German" },
  es: { countryCode: "spain", name: "Spanish" },
  ru: { countryCode: "russia", name: "Russian" },
  ko: { countryCode: "south_korea", name: "Korean" },
  fr: { countryCode: "france", name: "French" },
  ja: { countryCode: "japan", name: "Japanese" },
  pt: { countryCode: "portugal", name: "Portuguese" },
  tr: { countryCode: "turkey", name: "Turkish" },
  pl: { countryCode: "poland", name: "Polish" },
  ca: { countryCode: "spain", name: "Catalan" },
  nl: { countryCode: "netherlands", name: "Dutch" },
  ar: { countryCode: "saudi_arabia", name: "Arabic" },
  sv: { countryCode: "sweden", name: "Swedish" },
  it: { countryCode: "italy", name: "Italian" },
  id: { countryCode: "indonesia", name: "Indonesian" },
  hi: { countryCode: "india", name: "Hindi" },
  fi: { countryCode: "finland", name: "Finnish" },
  vi: { countryCode: "vietnam", name: "Vietnamese" },
  he: { countryCode: "israel", name: "Hebrew" },
  uk: { countryCode: "ukraine", name: "Ukrainian" },
  el: { countryCode: "greece", name: "Greek" },
  ms: { countryCode: "malaysia", name: "Malay" },
  cs: { countryCode: "czech_republic", name: "Czech" },
  ro: { countryCode: "romania", name: "Romanian" },
  da: { countryCode: "denmark", name: "Danish" },
  hu: { countryCode: "hungary", name: "Hungarian" },
  ta: { countryCode: "india", name: "Tamil" },
  no: { countryCode: "norway", name: "Norwegian" },
  th: { countryCode: "thailand", name: "Thai" },
  ur: { countryCode: "pakistan", name: "Urdu" },
  hr: { countryCode: "croatia", name: "Croatian" },
  bg: { countryCode: "bulgaria", name: "Bulgarian" },
  lt: { countryCode: "lithuania", name: "Lithuanian" },
  la: { countryCode: "vatican_city", name: "Latin" },
  mi: { countryCode: "new_zealand", name: "Maori" },
  ml: { countryCode: "india", name: "Malayalam" },
  cy: { countryCode: "wales", name: "Welsh" },
  sk: { countryCode: "slovakia", name: "Slovak" },
  te: { countryCode: "india", name: "Telugu" },
  fa: { countryCode: "iran", name: "Persian" },
  lv: { countryCode: "latvia", name: "Latvian" },
  bn: { countryCode: "bangladesh", name: "Bengali" },
  sr: { countryCode: "serbia", name: "Serbian" },
  az: { countryCode: "azerbaijan", name: "Azerbaijani" },
  sl: { countryCode: "slovenia", name: "Slovenian" },
  kn: { countryCode: "india", name: "Kannada" },
  et: { countryCode: "estonia", name: "Estonian" },
  mk: { countryCode: "north_macedonia", name: "Macedonian" },
  br: { countryCode: "france", name: "Breton" },
  eu: { countryCode: "spain", name: "Basque" },
  is: { countryCode: "iceland", name: "Icelandic" },
  hy: { countryCode: "armenia", name: "Armenian" },
  ne: { countryCode: "nepal", name: "Nepali" },
  mn: { countryCode: "mongolia", name: "Mongolian" },
  bs: { countryCode: "bosnia_and_herzegovina", name: "Bosnian" },
  kk: { countryCode: "kazakhstan", name: "Kazakh" },
  sq: { countryCode: "albania", name: "Albanian" },
  sw: { countryCode: "tanzania", name: "Swahili" },
  gl: { countryCode: "spain", name: "Galician" },
  mr: { countryCode: "india", name: "Marathi" },
  pa: { countryCode: "india", name: "Punjabi" },
  si: { countryCode: "sri_lanka", name: "Sinhala" },
  km: { countryCode: "cambodia", name: "Khmer" },
  sn: { countryCode: "zimbabwe", name: "Shona" },
  yo: { countryCode: "nigeria", name: "Yoruba" },
  so: { countryCode: "somalia", name: "Somali" },
  af: { countryCode: "south_africa", name: "Afrikaans" },
  oc: { countryCode: "france", name: "Occitan" },
  ka: { countryCode: "georgia", name: "Georgian" },
  be: { countryCode: "belarus", name: "Belarusian" },
  tg: { countryCode: "tajikistan", name: "Tajik" },
  sd: { countryCode: "pakistan", name: "Sindhi" },
  gu: { countryCode: "india", name: "Gujarati" },
  am: { countryCode: "ethiopia", name: "Amharic" },
  yi: { countryCode: "israel", name: "Yiddish" },
  lo: { countryCode: "laos", name: "Lao" },
  uz: { countryCode: "uzbekistan", name: "Uzbek" },
  fo: { countryCode: "denmark", name: "Faroese" },
  ht: { countryCode: "haiti", name: "Haitian Creole" },
  ps: { countryCode: "afghanistan", name: "Pashto" },
  tk: { countryCode: "turkmenistan", name: "Turkmen" },
  nn: { countryCode: "norway", name: "Norwegian Nynorsk" },
  mt: { countryCode: "malta", name: "Maltese" },
  sa: { countryCode: "india", name: "Sanskrit" },
  lb: { countryCode: "luxembourg", name: "Luxembourgish" },
  my: { countryCode: "myanmar", name: "Burmese" },
  bo: { countryCode: "tibet", name: "Tibetan" },
  tl: { countryCode: "philippines", name: "Tagalog" },
  mg: { countryCode: "madagascar", name: "Malagasy" },
  as: { countryCode: "india", name: "Assamese" },
  tt: { countryCode: "russia", name: "Tatar" },
  haw: { countryCode: "hawaii", name: "Hawaiian" },
  ln: { countryCode: "democratic_republic_of_congo", name: "Lingala" },
  ha: { countryCode: "nigeria", name: "Hausa" },
  ba: { countryCode: "russia", name: "Bashkir" },
  jw: { countryCode: "indonesia", name: "Javanese" },
  su: { countryCode: "indonesia", name: "Sundanese" },
};

const allLanguages = Object.keys(languageData).map((code) => ({
  code,
  name: languageData[code].name,
}));

export const SupportedLanguages = {
  get all(): SpokenLanguage[] {
    return allLanguages;
  },

  get defaultLanguage(): SpokenLanguage {
    return { code: "en", name: "English" };
  },

  /** Returns languages filtered to those supported by the given language codes. */
  forCodes(codes: string[] | undefined): SpokenLanguage[] {
    if (!codes) return allLanguages;
    const codeSet = new Set(codes);
    return allLanguages.filter((lang) => codeSet.has(lang.code));
  },

  /** Checks whether a language code is in the given supported set. */
  isSupported(code: string, supportedCodes: string[] | undefined): boolean {
    if (!supportedCodes) return true;
    return supportedCodes.includes(code);
  },

  findByCode(code: string): SpokenLanguage | null {
    const normalizedCode = code.toLowerCase();
    return languageData[normalizedCode]
      ? { code: normalizedCode, name: languageData[normalizedCode].name }
      : null;
  },

  countryCodeFor(code: string): string {
    return languageData[code]?.countryCode ?? "united_states";
  },

  transcribeOptionsFor(code: string): TranscribeOptions {
    const info = languageData[code];
    if (!info) {
      return { language: code };
    }
    return {
      language: info.whisperLanguage ?? code,
      prompt: info.prompt,
    };
  },
};

export const getCountryCode = (language: SpokenLanguage): string => {
  return SupportedLanguages.countryCodeFor(language.code);
};
